### PR TITLE
Allow partial retries in queued retry processor

### LINKED ISSUE #990: Allow partial retries of trace data in queued retry processor
The [queued retry processor](https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor#queued-retry-processor) retries the whole batch, however some exporters (e.g. Elasticsearch bulk API) can fail only part of the request (e.g. only specific spans).

A possible non-intrusive solution could be that the returned (exporter) error would carry properties that would identify objects that should be retried.

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <ploffay@redhat.com>

**Description:** <Describe what has changed. 

This patch allows partial retries in `queued_retry` processor. The following consumer after the retry processor initializes partial retry by returning an error type that encapsulates data that should be retried. When a different error type is returned the processor retries the original data set.

This PR does not contain any breaking changes nor change of the current behavior of the `queued_retry` processor.

**Link to tracking Issue:** <Issue number if applicable>

Resolves #990

**Testing:** < Describe what testing was performed and which tests were added.>

**Documentation:** < Describe the documentation added.>

Not needed only godoc.

### CODE REVIEW

**Comment by owais** [comment]
Does this mean exporters need to know about the existence of queued processor and may be use it as a dependency (to import the error type)? Can we may be move the error type to the components or some other core package and generalize it so it's not highly specific to queued retry? We could rename it to something generic and also rename the `toRetry` field to something like `failures` or `failed`. We'd also need different error types for metrics, traces and logs or may be we can have `failedMetrics`, `failedTraces` fields on the same error? Once we have this, exporters can freely chose to use return this error and any component in the pipeline before can add support to handle partial failures. This would enable people to implement custom error handling/retrying processors that can also support partial retries.

**Comment by pavolloffay** [comment]
> Does this mean exporters need to know about the existence of queued processor and may be use it as a dependency

correct, currently exporters would have to import it if they want to use the error type.

> We'd also need different error types for metrics, traces and logs or may be we can have failedMetrics, failedTraces fields on the same error?

A different error type would work, or a single type with multiple factory methods and getters for data types.

**Comment by pavolloffay** [comment]
I have moved the error to the `consumer` package and made it more generic.

**Comment by bogdandrutu** [comment]
Please rebase I made some refactoring in the qretry

**Comment by pavolloffay** [comment]
Rebased

File: processor/queuedprocessor/queued_processor.go [hunk 1]
```diff
@@ -174,6 +174,9 @@ func (sp *queuedSpanProcessor) processItemFromQueue(item *queueItem) {
 		// throw away the batch
 		sp.onItemDropped(item, err)
 		return
+	} else if partialErr, isPartial := err.(consumererror.PartialError); isPartial {
```
Comment (bogdandrutu) [inline processor/queuedprocessor/queued_processor.go:177]: Don't need `else if` because the previous returns. just an `if`

Comment (bogdandrutu) [reply processor/queuedprocessor/queued_processor.go:177]: Will fix this in my metrics PR.

**Pull request merged**